### ISSUE TITLE
Client object breaks trying to call method

### ISSUE THREAD

**Issue by PyMap** [vgrem/Office365-REST-Python-Client#397]
Hi all! Working with onedrive API I found out that, once I instantiate the `client = GraphClient(acquire_token_func(tenant_id, client_id, client_secret))`:

![image](https://user-images.githubusercontent.com/44501803/130526161-851e59c9-1ce8-47e4-b870-2d904bebe16c.png)

I think that the error happens here https://github.com/vgrem/Office365-REST-Python-Client/blob/master/office365/graph_client.py#L94

... because the `_acquire_token_callback()` shouldn't be called as a method. At least, when I explore this:
`client._acquire_token_callback['access_token']`

 the access token shows up correctly (which make me infer that client id, secret and tenant are good

**Comment by vgrem** [comment]
Greetings!

`GraphClient `class constructor accept a _callback_ function as an input argument, for example:

`
client = GraphClient(acquire_token_by_client_credentials)
`

where

```python
def acquire_token_by_client_credentials():
    authority_url = 'https://login.microsoftonline.com/{0}'.format(tenant_name)
    app = msal.ConfidentialClientApplication(
        authority=authority_url,
        client_id=client_id,
        client_credential=client_secret
    )
    return app.acquire_token_for_client(scopes=["https://graph.microsoft.com/.default"])

```


The _callback function_ is expected to return a `dict `representing the json token (which should contain `access_token` key):

```json
{
   'token_type': 'Bearer', 
   'scope': 'profile openid email https://graph.microsoft.com/Mail.Send', 
   'access_token': '--access token goes here--', 
}

```

By default [Microsoft Authentication Library (MSAL) for Python](https://github.com/AzureAD/microsoft-authentication-library-for-python) library is utilized to acquire tokens.

And finally, callback function gets invoked call during `authenticate_request `call by library.

**Comment by PyMap** [comment]
@vgrem thanks for the explanation! The `client` object I'm showing in the first post was instantiated in the same way do. I mean, by calling the acquire_token_by_client_credentials method with my tenant, client id and secret.

When I try to call drive items from the client object, like this
```
drives = client.drives.get().execute_query()
```
... then the error I showed in the post raises up. 

I manage to get my token if I do the following:
```
client._acquire_token_callback
```

This returns me something like:
```
{'access_token': 'my_token',
 'expires_in': 86399,
 'ext_expires_in': 86399,
 'token_type': 'Bearer'}
```

In the description, the error shows that method breaks when trying to call the callback function. But if I try it out of the client.drives.get() (like I showed you above) it works fine:

![image](https://user-images.githubusercontent.com/44501803/130626128-6bdd2e38-b592-4d4c-9b49-b30dcc80d5cb.png)

To sum up, in my case this is breaking in line 94 of the graph_client.py when trying to do `token_json = self._acquire_token_callback()`. That's why I'm asking if it should be `_acquire_token_callback` instead.

I'm not managing to instantiate the drives object to call the .web_url attribute

**Comment by not1q84-1** [comment]
Hi PyMap, 

I have started working with vgrem's package only yesterday and following his examples worked fine for me. But then I restructured and integrated the code in my existing code - and accidentally initiated the client instances by calling the acquire_token_func using "() "at the end: client = GraphClient(acquire_token_func()) - this produced exactly the error you  are describing here also for me... did you iniate the client perhaps also with "()"

**Comment by jgentil** [comment]
> The _callback function_ is expected to return a `dict `representing the json token (which should contain `access_token` key):


@vgrem If this is true then why does it access it by property?

https://github.com/vgrem/Office365-REST-Python-Client/blob/cb4b7ce7cbb9e7e951dacc5e4f38a31c045038cd/office365/runtime/auth/providers/oauth_token_provider.py#L20-L22


As you can see it tries to access the value returned from the function token as an object (self._cached_token.tokenType`), not as a dict.  

Please re-open this issue.

**Comment by jgentil** [comment]
The callback function needs to return an instance of `office365.runtime.auth.token_response.TokenResponse` not a dict.

```python
    token = app.acquire_token_for_client(scopes=["https://graph.microsoft.com/.default"])
    return TokenResponse(**token)
```
works for me.

**Comment by vgrem** [comment]
In 2026, the built-in auth builders are cleaner and avoid this confusion entirely:

```python
# recommended — no custom acquire_token function needed
client = GraphClient(tenant).with_client_secret(client_id, client_secret)
client = GraphClient(tenant).with_client_certificate(client_id, thumbprint, private_key)
client = GraphClient(tenant).with_token_interactive(client_id)
```

If you need a custom acquire function (e.g. existing MSAL app), the function must return the MSAL token dict directly - `MSAL `already returns the correct format:

```python
def acquire_token():
    app = msal.ConfidentialClientApplication(
        client_id, authority=f"https://login.microsoftonline.com/{tenant}",
        client_credential=client_secret
    )
    return app.acquire_token_for_client(["https://graph.microsoft.com/.default"])

client = GraphClient(acquire_token)  # ← function reference, no ()
```

The `TokenResponse(**token)` workaround suggested in the thread is no longer needed - `TokenResponse` now accepts the raw `MSAL` dict directly
